### PR TITLE
EZP-28250: As an editor, I want to edit a RelationList field type

### PR DESCRIPTION
**Pull request by sunpietro**
| Question      | Answer
| ------------- | ---
| Tickets       | https://jira.ez.no/browse/EZP-28250
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Tests pass?   | yes/no
| Doc needed?   | yes
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

**Sub tasks:**
- [x] https://jira.ez.no/browse/EZP-28396 - Implement the fieldtype UI,

**Requires:**
 - https://github.com/ezsystems/ezpublish-kernel/pull/2178
 - https://github.com/ezsystems/repository-forms/pull/183
 - https://github.com/ezsystems/ezplatform-admin-ui-modules/pull/26

### CODE REVIEW

File: src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        });
+        const udwContainer = document.getElementById('react-udw');
+        const token = document.querySelector('meta[name="CSRF-Token"]').content;
+        const siteaccess = document.querySelector('meta[name="SiteAccess"]').content;
+        const sourceInput = fieldContainer.querySelector(SELECTOR_INPUT);
+        const relationsContainer = fieldContainer.querySelector('.ez-relations__list');
+        const relationsWrapper = fieldContainer.querySelector('.ez-relations__wrapper');
+        const relationsCTA = fieldContainer.querySelector('.ez-relations__cta');
+        const addBtn = fieldContainer.querySelector(SELECTOR_BTN_ADD)
+        const selectedItemsLimit = 2;
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js]: hardcode 2?

File: src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            ]
+        });
+        const udwContainer = document.getElementById('react-udw');
+        const token = document.querySelector('meta[name="CSRF-Token"]').content;
+        const siteaccess = document.querySelector('meta[name="SiteAccess"]').content;
+        const sourceInput = fieldContainer.querySelector(SELECTOR_INPUT);
+        const relationsContainer = fieldContainer.querySelector('.ez-relations__list');
+        const relationsWrapper = fieldContainer.querySelector('.ez-relations__wrapper');
+        const relationsCTA = fieldContainer.querySelector('.ez-relations__cta');
+        const addBtn = fieldContainer.querySelector(SELECTOR_BTN_ADD)
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js]: missing `;`

File: src/bundle/Resources/views/fieldtypes/ezobjectrelationlist.html.twig [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            </tbody>
+        </table>
+    </div>
+    <div class="ez-relations__cta" {% if has_relations %}hidden="true"{% endif %}>
+        <p class="ez-relations__cta-text">Set up multiple relations with up to {{ limit }} content items</p>
+        <button class="ez-relations__cta-btn btn btn-primary">
+            <svg class="ez-icon">
+                <use xlink:href="/bundles/ezplatformadminui/img/ez-icons.svg#relations"></use>
+            </svg>
+            <span class="ez-relations__cta-btn-label">Select content</span>
```
Comment (dew326) [inline src/bundle/Resources/views/fieldtypes/ezobjectrelationlist.html.twig]: translation

File: src/bundle/Resources/views/fieldtypes/ezobjectrelationlist.html.twig [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                        <td>{{ relation.contentType.name }}</td>
+                        <td>{{ relation.contentInfo.publishedDate|localizeddate('short', 'medium') }}</td>
+                        <td colspan="2"><input class="ez-relations__order-input" type="number" value="{{ loop.index }}" /></td>
+                    </tr>
+                {% endfor %}
+            </tbody>
+        </table>
+    </div>
+    <div class="ez-relations__cta" {% if has_relations %}hidden="true"{% endif %}>
+        <p class="ez-relations__cta-text">Set up multiple relations with up to {{ limit }} content items</p>
```
Comment (dew326) [inline src/bundle/Resources/views/fieldtypes/ezobjectrelationlist.html.twig]: translation

**Comment by sunpietro** [comment]
ping @dew326

File: src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        const token = document.querySelector('meta[name="CSRF-Token"]').content;
+        const siteaccess = document.querySelector('meta[name="SiteAccess"]').content;
+        const sourceInput = fieldContainer.querySelector(SELECTOR_INPUT);
+        const relationsContainer = fieldContainer.querySelector('.ez-relations__list');
+        const relationsWrapper = fieldContainer.querySelector('.ez-relations__wrapper');
+        const relationsCTA = fieldContainer.querySelector('.ez-relations__cta');
+        const addBtn = fieldContainer.querySelector(SELECTOR_BTN_ADD);
+        const selectedItemsLimit = parseInt(relationsContainer.dataset.limit, 10);
+        const closeUDW = () => udwContainer.innerHTML = '';
+        const onCancel = () => closeUDW();
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js]: Do we need this? Couldn't you just set in UDW: `onCancel: closeUDW`?

File: src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        };
+        const openUDW = (event) => {
+            event.preventDefault();
+
+            ReactDOM.render(React.createElement(global.eZ.modules.UniversalDiscovery, {
+                onConfirm,
+                onCancel,
+                confirmLabel: 'Confirm selection',
+                title: 'Select content',
+                multiple: true,
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js]: What do you think to set `multiple: false` when `selectedItemsLimit === 1`?

File: src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js (hunk 5 above)
Comment (sunpietro) [reply src/bundle/Resources/public/js/scripts/fieldType/ezobjectrelationlist.js]: That's another option. Good point! :)

**Comment by mnocon** [comment]
@sunpietro 
I have a JS error when trying to publish a Content Item with `ezobjectrelation` (I'm using ezplatform, not ezplatform-ee). 

Steps:
1. Login as admin
2. Start creating a new Article
3. Fill title, e-mail and Intro fields
4. Click "Publish"

Result:
Item is not published. Errors in console:
```
Uncaught TypeError: Cannot read property 'querySelector' of null
    at EzObjectRelationListValidator.validateInput (fc7891e.js:5084)
    at EzObjectRelationListValidator.validateField (fc7891e.js:3058)
```
and
```
Uncaught TypeError: Cannot read property 'querySelector' of null
    at EzObjectRelationListValidator.validateInput (fc7891e.js:5084)
    at EzObjectRelationListValidator.validateField (fc7891e.js:3058)
    at fieldsToValidate.forEach.field (fc7891e.js:3121)
    at Array.forEach (<anonymous>)
    at EzObjectRelationListValidator.isValid (fc7891e.js:3119)
    at validators.map.validator (fc7891e.js:2878)
    at Array.map (<anonymous>)
    at HTMLButtonElement.clickHandler (fc7891e.js:2878)
    at HTMLButtonElement.trigger (0bfb931.js:952)
```
Also, the translations are not displayed in "prod" mode:
screen shot 2017-12-11 at 13 22 56

**Pull request merged**